src/transformers/modeling_utils.py

```diff
@@ -2609,6 +2642,25 @@ def save_pretrained(
 
         # Save the model
         for shard_file, shard in shards.items():
+            # remake shard with onloaded parameters if necessary
+            if module_map:
+                if accelerate_version < version.parse("0.31"):
+                    raise ImportError(
+                        f"You need accelerate version to be greater or equal than 0.31 to save models with offloaded parameters. Detected version {accelerate_version}. "
+                        f"Please upgrade accelerate with `pip install -U accelerate`"
+                    )
+                # init state_dict for this shard
+                state_dict = {name: "" for name in shard}
+                for module_name in state_dict.keys():
```

Comment:
Could we instead iterate over the names in `shard`? This way we're not iterating over an object which we're modifying 

```suggestion
                for module_name in shard:
```